rust/sbp/src/codec.rs

```diff
@@ -0,0 +1,405 @@
+#[cfg(feature = "json")]
+pub use json::{json2json, json2sbp, sbp2json};
+
+pub(crate) mod sbp {
+    use bytes::{Buf, BufMut, BytesMut};
+    use futures::stream::Stream;
+    use tokio::io::{AsyncRead, AsyncWrite};
+    use tokio_util::codec::{Decoder, Encoder, FramedRead, FramedWrite};
+
+    use crate::{
+        messages::{SBPMessage, SBP},
+        parser::{parse_sbp, ParseResult},
+        Error, Result,
+    };
+
+    const MAX_FRAME_LENGTH: usize =
+        crate::MSG_HEADER_LEN + crate::SBP_MAX_PAYLOAD_SIZE + crate::MSG_CRC_LEN;
+
+    pub fn stream_messages<R: AsyncRead>(input: R) -> impl Stream<Item = Result<SBP>> {
+        SbpDecoder::decode_reader(input)
+    }
+
+    pub struct SbpDecoder {}
+
+    impl SbpDecoder {
+        pub fn decode_reader<R: AsyncRead>(input: R) -> FramedRead<R, SbpDecoder> {
+            FramedRead::new(input, SbpDecoder {})
+        }
+    }
+
+    impl Decoder for SbpDecoder {
+        type Item = SBP;
+        type Error = Error;
+
+        fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>> {
+            match parse_sbp(&src) {
+                ParseResult::Ok((bytes_read, msg)) => {
+                    src.advance(bytes_read);
+                    log::trace!("{:?}", msg);
+                    Ok(Some(msg))
+                }
+                ParseResult::Err((bytes_read, err)) => {
+                    src.advance(bytes_read);
+                    log::warn!("{}", err);
+                    Ok(None)
+                }
+                ParseResult::Incomplete => {
+                    src.reserve(MAX_FRAME_LENGTH);
+                    Ok(None)
+                }
+            }
+        }
+
+        fn decode_eof(&mut self, buf: &mut BytesMut) -> Result<Option<Self::Item>> {
+            match self.decode(buf) {
+                Ok(Some(frame)) => Ok(Some(frame)),
+                _ => Ok(None),
+            }
+        }
+    }
+
+    pub struct SbpEncoder {
+        frame: Vec<u8>,
+    }
+
+    impl SbpEncoder {
+        pub fn encode_writer<W: AsyncWrite>(writer: W) -> FramedWrite<W, SbpEncoder> {
+            FramedWrite::new(
+                writer,
+                SbpEncoder {
+                    frame: Vec::with_capacity(MAX_FRAME_LENGTH),
+                },
+            )
+        }
+    }
+
+    impl Encoder<SBP> for SbpEncoder {
+        type Error = Error;
+
+        fn encode(&mut self, msg: SBP, dst: &mut BytesMut) -> Result<()> {
+            self.frame.clear();
+            match msg.write_frame(&mut self.frame) {
```

Comment:
If the methods on `SbpSerialize` took `&mut BytesMut` or maybe `&mut impl bytes::BufMut` we could write the message directly to `dst`